Broken nix build

**To Reproduce**
```shell
nix build github:emacs-ng/emacs-ng --option substituters "https://emacsng.cachix.org" --option trusted-public-keys "emacsng.cachix.org-1:i7wOr4YpdRpWWtShI8bT6V7lOTnPeI7Ho6HaZegFWMI=" -o emacsNg
```
output:
```shell
error: builder for '/nix/store/9ik0f6jnz1gfx66yglprrh3dga7r5h7j-emacs-ng-30.0.50.drv' failed with exit code 2;
       last 10 log lines:
       > *** "make all" failed with exit status 2.
       > ***
       > *** You could try to:
       > *** - run "make bootstrap", which might fix the problem
       > *** - run "make V=1", which displays the full commands invoked by make,
       > ***   to further investigate the problem
       > ***
       > make[1]: *** [Makefile:428: advice-on-failure] Error 2
       > make[1]: Leaving directory '/tmp/nix-build-emacs-ng-30.0.50.drv-0/hwr6w3a323wlp559k3a31si5a102nsqx-source'
       > make: *** [Makefile:384: all] Error 2
       For full logs, run 'nix log /nix/store/9ik0f6jnz1gfx66yglprrh3dga7r5h7j-emacs-ng-30.0.50.drv'.
```

> 

```shell
nix build github:emacs-ng/emacs-ng#emacsng --option substituters "https://emacsng.cachix.org" --option trusted-public-keys "emacsng.cachix.org-1:i7wOr4YpdRpWWtShI8bT6V7lOTnPeI7Ho6HaZegFWMI="
error: builder for '/nix/store/9ik0f6jnz1gfx66yglprrh3dga7r5h7j-emacs-ng-30.0.50.drv' failed with exit code 2;
       last 10 log lines:
       > *** "make all" failed with exit status 2.
       > ***
       > *** You could try to:
       > *** - run "make bootstrap", which might fix the problem
       > *** - run "make V=1", which displays the full commands invoked by make,
       > ***   to further investigate the problem
       > ***
       > make[1]: *** [Makefile:428: advice-on-failure] Error 2
       > make[1]: Leaving directory '/tmp/nix-build-emacs-ng-30.0.50.drv-0/hwr6w3a323wlp559k3a31si5a102nsqx-source'
       > make: *** [Makefile:384: all] Error 2
       For full logs, run 'nix log /nix/store/9ik0f6jnz1gfx66yglprrh3dga7r5h7j-emacs-ng-30.0.50.drv'.
```

--- a/Cargo.in
+++ b/Cargo.in
@@ -109,6 +109,7 @@ javascript = ["dep:js", "ng-bindgen/javascript"]
 ng-module = ["dep:ng_module", "ng-bindgen/ng-module"]
 # Enable glyphs debugging code.
 glyph-debug = ["wrterm?/capture"]
+profiling = ["wrterm?/profiling"]
 
 ### Patches
 

--- a/configure.ac
+++ b/configure.ac
@@ -689,6 +689,7 @@ AC_ARG_ENABLE([profiling],
 [ac_enable_profiling="${enableval}"],[])
 if test x$ac_enable_profiling != x ; then
    PROFILING_CFLAGS="-DPROFILING=1 -pg"
+   CARGO_CHECKING_FEATURES="${CARGO_CHECKING_FEATURES}\"profiling\", "
 else
    PROFILING_CFLAGS=
 fi

--- a/rust_src/crates/webrender/Cargo.toml
+++ b/rust_src/crates/webrender/Cargo.toml
@@ -99,6 +99,7 @@ wayland = [
   "glutin?/wayland"
 ]
 capture=["webrender/capture", "webrender/serialize_program"]
+profiling=["webrender/profiler"]
 sw_compositor=["webrender/sw_compositor"]
 pselect=["dep:nix"]
 tokio=["dep:tokio"]

--- a/rust_src/crates/webrender/src/event_loop.rs
+++ b/rust_src/crates/webrender/src/event_loop.rs
@@ -72,16 +72,13 @@ impl WrEventLoop {
                 let el = EventLoopBuilder::<i32>::with_user_event().build();
                 #[cfg(use_tao)]
                 let el = EventLoop::<i32>::with_user_event();
-
                 let clipboard = Clipboard::build(&el);
                 (el, clipboard)
             };
-
             Arc::new(Mutex::new(Self { clipboard, el }))
         })
     }
 }
-
 pub fn global_event_buffer() -> &'static Mutex<Vec<GUIEvent>> {
     static EVENT_BUFFER: OnceLock<Mutex<Vec<GUIEvent>>> = OnceLock::new();
     EVENT_BUFFER.get_or_init(|| Mutex::new(Vec::new()))

--- a/rust_src/crates/webrender/src/event_loop.rs
+++ b/rust_src/crates/webrender/src/event_loop.rs
@@ -18,7 +18,7 @@ use crate::window_system::api::{
     monitor::MonitorHandle,
     platform::run_return::EventLoopExtRunReturn,
 };
-use emacs::bindings::{inhibit_window_system, thread_select};
+use emacs::bindings::{inhibit_window_system, noninteractive, thread_select};
 use libc::{c_void, fd_set, pselect, sigset_t, timespec};
 
 pub type GUIEvent = Event<'static, i32>;
@@ -135,14 +135,12 @@ pub extern "C" fn winit_select(
     timeout: *mut timespec,
     _sigmask: *mut sigset_t,
 ) -> i32 {
-    log::trace!("winit select");
-    let lock_result = WrEventLoop::global().try_lock();
-
-    if lock_result.is_err() || unsafe { inhibit_window_system } {
-        if lock_result.is_err() {
-            log::debug!("Failed to grab a lock {:?}", lock_result.err());
-        }
-
+    log::trace!(
+        "winit select inhibit_window_system: {}, noninteractive: {}",
+        unsafe { inhibit_window_system },
+        unsafe { noninteractive }
+    );
+    if unsafe { inhibit_window_system || noninteractive } {
         return unsafe {
             thread_select(
                 Some(pselect),
@@ -156,6 +154,11 @@ pub extern "C" fn winit_select(
         };
     }
 
+    let lock_result = WrEventLoop::global().try_lock();
+    if lock_result.is_err() {
+        log::error!("Failed to grab a lock {:?}", lock_result.err());
+        return -1;
+    }
     let mut event_loop = lock_result.unwrap();
 
     handle_select(

--- a/rust_src/crates/webrender/src/event_loop.rs
+++ b/rust_src/crates/webrender/src/event_loop.rs
@@ -78,6 +78,9 @@ impl WrEventLoop {
             Arc::new(Mutex::new(Self { clipboard, el }))
         })
     }
+    pub fn get() -> Option<&'static Arc<Mutex<WrEventLoop>>> {
+        EVENT_LOOP.get()
+    }
 }
 pub fn global_event_buffer() -> &'static Mutex<Vec<GUIEvent>> {
     static EVENT_BUFFER: OnceLock<Mutex<Vec<GUIEvent>>> = OnceLock::new();

--- a/rust_src/crates/webrender/src/image.rs
+++ b/rust_src/crates/webrender/src/image.rs
@@ -58,18 +58,8 @@ impl ImageExt for ImageRef {
         let specified_data = self.specified_data();
         let specified_file = self.specified_file();
         if specified_data.is_nil() {
-            let file: LispObject =
-                unsafe { emacs::bindings::image_find_image_file(specified_file) };
-            if !file.is_string() || file.is_nil() {
-                image_error!("Cannot find image file {:?}", specified_file);
-                return None;
-            }
-            return Some(ImageSource::File(file));
+            return Some(ImageSource::File(specified_file));
         } else {
-            if !specified_data.is_string() {
-                image_error!("Invalid image data {:?}", specified_data);
-                return None;
-            }
             return Some(ImageSource::Data(specified_data));
         }
     }

--- a/rust_src/crates/webrender/src/image_cache.rs
+++ b/rust_src/crates/webrender/src/image_cache.rs
@@ -359,18 +359,16 @@ impl ImageSource {
 /// TODO We plan implement spec(rotate/scale) using WebRender transform
 pub struct ImageCache(FastHashMap<ImageHash, ImageCacheResult>);
 
-static mut IMAGE_CACHE: OnceLock<Arc<Mutex<ImageCache>>> = OnceLock::new();
+static IMAGE_CACHE: OnceLock<Arc<Mutex<ImageCache>>> = OnceLock::new();
 impl ImageCache {
     pub fn new() -> Arc<Mutex<Self>> {
         Arc::new(Mutex::new(Self(FastHashMap::default())))
     }
     pub fn global() -> &'static Arc<Mutex<ImageCache>> {
-        unsafe {
-            IMAGE_CACHE.get_or_init(|| {
-                log::trace!("image cache is being created...");
-                Self::new()
-            })
-        }
+        IMAGE_CACHE.get_or_init(|| {
+            log::trace!("image cache is being created...");
+            Self::new()
+        })
     }
     fn cache_mut<P, T>(p: P) -> Option<T>
     where

--- a/rust_src/crates/webrender/src/image_cache.rs
+++ b/rust_src/crates/webrender/src/image_cache.rs
@@ -4,6 +4,7 @@ use super::image::ImageRef;
 use super::image::WrPixmapRef;
 use crate::font_db::FontDB;
 use emacs::bindings::Fequal;
+use emacs::bindings::Fmake_temp_file_internal;
 use emacs::bindings::EMACS_UINT;
 use emacs::globals::Qnil;
 use emacs::globals::Qsvg;
@@ -154,16 +155,25 @@ impl ImageSource {
     {
         match self {
             Self::File(file) => {
+                let file: LispObject = unsafe { emacs::bindings::image_find_image_file(file) };
+                if !file.is_string() || file.is_nil() {
+                    image_error!("Cannot find image file: {:?}", file);
+                    return None;
+                }
                 let filename = String::from(file);
                 match std::fs::read(&filename) {
                     Ok(result) => p(result.as_slice()),
                     Err(e) => {
-                        image_error!("Error open image file {:?} {e:?}", file);
+                        image_error!("Error open image file: {:?} {e:?}", file);
                         return None;
                     }
                 }
             }
             Self::Data(data) => {
+                if !data.is_string() {
+                    image_error!("Invalid image data: {:?}", data);
+                    return None;
+                }
                 let data = data.as_string().unwrap();
                 p(data.as_slice())
             }
@@ -193,7 +203,18 @@ impl ImageSource {
             match Self::svg_to_png(data, &opt) {
                 Some(bytes) => p(bytes.as_slice()),
                 None => {
-                    image_error!("Error reading svg file {:?}", self);
+                    let file = match self {
+                        Self::File(file) => file,
+                        Self::Data(data) => unsafe {
+                            Fmake_temp_file_internal(
+                                "invalid_resvg".into(),
+                                Qnil,
+                                ".svg".into(),
+                                data,
+                            )
+                        },
+                    };
+                    image_error!("Error reading svg file: {:?}", file);
                     return None;
                 }
             }
@@ -212,7 +233,7 @@ impl ImageSource {
                 match result {
                     Ok(data) => Some(data),
                     Err(e) => {
-                        image_error!("Error decoding image {:?} {e:?}", def);
+                        image_error!("Error decoding image: {:?} {e:?}", def);
                         return None;
                     }
                 }
@@ -238,7 +259,7 @@ impl ImageSource {
         let mut tree = match result {
             Ok(result) => result,
             Err(e) => {
-                image_error!("Failed to parse svg {e:?}");
+                image_error!("Failed to parse svg: {e:?}");
                 return None;
             }
         };
@@ -452,7 +473,7 @@ impl ImageCache {
         let source = image.source();
 
         if source.is_none() {
-            image_error!("Invalid image source {:?}", image.spec());
+            image_error!("Invalid image source: {:?}", image.spec());
             return;
         }
 

--- a/rust_src/crates/webrender/src/lib.rs
+++ b/rust_src/crates/webrender/src/lib.rs
@@ -159,11 +159,8 @@ pub use crate::platform::macos;
 
 pub use crate::color::*;
 pub use crate::font::*;
-pub use crate::term::*;
 #[cfg(window_system_winit)]
 pub use crate::window_system::term::*;
-#[cfg(window_system_winit)]
-pub use crate::wrterm::*;
 
 #[cfg(not(test))]
 #[cfg(window_system_winit)]

--- a/rust_src/crates/webrender/src/window_system/winit/term.rs
+++ b/rust_src/crates/webrender/src/window_system/winit/term.rs
@@ -55,10 +55,8 @@ use emacs::{
     frame::{all_frames, LispFrameRef, Lisp_Frame},
     globals::{Qnil, Qwinit},
     keyboard::allocate_keyboard,
-    lisp::{ExternalPtr, LispObject},
+    lisp::LispObject,
 };
-
-pub type TerminalRef = ExternalPtr<terminal>;
 
 fn get_frame_parm_handlers() -> [frame_parm_handler; 48] {
     // Keep this list in the same order as frame_parms in frame.c.

--- a/rust_src/crates/webrender/src/wrterm.rs
+++ b/rust_src/crates/webrender/src/wrterm.rs
@@ -609,7 +609,11 @@ pub fn webrender_monitor_to_emacs_monitor(m: MonitorHandle) -> (MonitorInfo, Opt
 /// Internal use only, use `display-monitor-attributes-list' instead.
 #[lisp_fn(min = "0")]
 pub fn wr_display_monitor_attributes_list(_terminal: LispObject) -> LispObject {
-    let event_loop = WrEventLoop::global().try_lock().unwrap();
+    let event_loop = WrEventLoop::get();
+    if event_loop.is_none() {
+        return Qnil;
+    }
+    let event_loop = event_loop.unwrap().try_lock().unwrap();
 
     let monitors: Vec<_> = event_loop.get_available_monitors().collect();
     let primary_monitor = event_loop.get_primary_monitor();
@@ -676,7 +680,11 @@ pub fn wr_display_monitor_attributes_list(_terminal: LispObject) -> LispObject {
 /// each physical monitor, use `display-monitor-attributes-list'.
 #[lisp_fn(min = "0")]
 pub fn x_display_pixel_width(_terminal: LispObject) -> LispObject {
-    let event_loop = WrEventLoop::global().try_lock().unwrap();
+    let event_loop = WrEventLoop::get();
+    if event_loop.is_none() {
+        return Qnil;
+    }
+    let event_loop = event_loop.unwrap().try_lock().unwrap();
 
     let primary_monitor = event_loop.get_primary_monitor();
 
@@ -699,7 +707,11 @@ pub fn x_display_pixel_width(_terminal: LispObject) -> LispObject {
 /// each physical monitor, use `display-monitor-attributes-list'.
 #[lisp_fn(min = "0")]
 pub fn x_display_pixel_height(_terminal: LispObject) -> LispObject {
-    let event_loop = WrEventLoop::global().try_lock().unwrap();
+    let event_loop = WrEventLoop::get();
+    if event_loop.is_none() {
+        return Qnil;
+    }
+    let event_loop = event_loop.unwrap().try_lock().unwrap();
 
     let primary_monitor = event_loop.get_primary_monitor();
 
@@ -727,7 +739,11 @@ pub fn x_own_selection_internal(
     value: LispObject,
     _frame: LispObject,
 ) -> LispObject {
-    let mut event_loop = WrEventLoop::global().try_lock().unwrap();
+    let event_loop = WrEventLoop::get();
+    if event_loop.is_none() {
+        return Qnil;
+    }
+    let mut event_loop = event_loop.unwrap().try_lock().unwrap();
 
     let clipboard = event_loop.get_clipboard();
 
@@ -758,7 +774,11 @@ pub fn x_get_selection_internal(
     _time_stamp: LispObject,
     _terminal: LispObject,
 ) -> LispObject {
-    let mut event_loop = WrEventLoop::global().try_lock().unwrap();
+    let event_loop = WrEventLoop::get();
+    if event_loop.is_none() {
+        return Qnil;
+    }
+    let mut event_loop = event_loop.unwrap().try_lock().unwrap();
 
     let clipboard = event_loop.get_clipboard();
     let contents: &str = &clipboard.read();

